compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts

```diff
@@ -0,0 +1,517 @@
+      }
+    }
+
+    nodes.set(block.id, {
+      block,
+      assumedNonNullObjects,
+    });
+  }
+  return nodes;
+}
+
+function deriveNonNull(fn: HIRFunction, nodes: Map<BlockId, BlockInfo>): void {
+  const succ = new Map<BlockId, Set<BlockId>>();
+  const terminalPreds = new Set<BlockId>();
+
+  for (const [blockId, block] of fn.body.blocks) {
+    for (const pred of block.preds) {
+      getOrInsertDefault(succ, pred, new Set()).add(blockId);
+    }
+    if (block.terminal.kind === 'throw' || block.terminal.kind === 'return') {
+      terminalPreds.add(blockId);
+    }
+  }
+
+  function recursivelyDeriveNonNull(
+    nodeId: BlockId,
+    direction: 'forward' | 'backward',
+    traversalState: Map<BlockId, 'active' | 'done'>,
+    nonNullObjectsByBlock: Map<BlockId, Set<PropertyLoadNode>>,
+  ): boolean {
+    /**
+     * Avoid re-visiting computed or currently active nodes, which can
+     * occur when the control flow graph has backedges.
+     */
+    if (traversalState.has(nodeId)) {
+      return false;
+    }
+    traversalState.set(nodeId, 'active');
+
+    const node = nodes.get(nodeId);
+    if (node == null) {
+      CompilerError.invariant(false, {
+        reason: `Bad node ${nodeId}, kind: ${direction}`,
+        loc: GeneratedSource,
+      });
+    }
+    const neighbors = Array.from(
+      direction === 'backward' ? (succ.get(nodeId) ?? []) : node.block.preds,
+    );
+
+    let changed = false;
+    for (const pred of neighbors) {
+      if (!traversalState.has(pred)) {
+        const neighborChanged = recursivelyDeriveNonNull(
+          pred,
+          direction,
+          traversalState,
+          nonNullObjectsByBlock,
+        );
+        changed ||= neighborChanged;
+      }
+    }
+    /**
+     * Active neighbors can be filtered out as we're solving for the following
+     * relation.
+     * X = Intersect(X_neighbors, X)
+     * Non-active neighbors with no recorded results can occur due to backedges.
+     * it's not safe to assume they can be filtered out (e.g. not intersected)
+     */
+    const neighborAccesses = Set_intersect([
+      ...(Array.from(neighbors)
+        .filter(n => traversalState.get(n) === 'done')
+        .map(n => nonNullObjectsByBlock.get(n) ?? new Set()) as Array<
+        Set<PropertyLoadNode>
+      >),
+    ]);
+
+    const prevSize = nonNullObjectsByBlock.get(nodeId)?.size;
+    nonNullObjectsByBlock.set(
+      nodeId,
+      Set_union(node.assumedNonNullObjects, neighborAccesses),
+    );
+    traversalState.set(nodeId, 'done');
+
+    changed ||= prevSize !== nonNullObjectsByBlock.get(nodeId)!.size;
+    CompilerError.invariant(
+      prevSize == null || prevSize <= nonNullObjectsByBlock.get(nodeId)!.size,
+      {
+        reason: '[CollectHoistablePropertyLoads] Nodes shrank!',
+        description: `${nodeId} ${direction} ${prevSize} ${
+          nonNullObjectsByBlock.get(nodeId)!.size
+        }`,
+        loc: GeneratedSource,
+      },
+    );
+    return changed;
```

Comment:
Will remove the fixed-point iteration logic in a later PR after more testing internally